src/cargo/util/flock.rs

```diff
@@ -0,0 +1,276 @@
+            opts: &OpenOptions,
+            state: State,
+            config: &Config,
+            msg: &str) -> CargoResult<FileLock> {
+        let path = self.root.join(path);
+
+        // If we want an exclusive lock then if we fail because of NotFound it's
+        // likely because an intermediate directory didn't exist, so try to
+        // create the directory and then continue.
+        let f = try!(opts.open(&path).or_else(|e| {
+            if e.kind() == io::ErrorKind::NotFound && state == State::Exclusive {
+                try!(create_dir_all(path.parent().unwrap()));
+                opts.open(&path)
+            } else {
+                Err(e)
+            }
+        }).chain_error(|| {
+            human(format!("failed to open: {}", path.display()))
+        }));
+        match state {
+            State::Exclusive => {
+                try!(acquire(config, msg, &path,
+                             &|| f.try_lock_exclusive(),
+                             &|| f.lock_exclusive()));
+            }
+            State::Shared => {
+                try!(acquire(config, msg, &path,
+                             &|| f.try_lock_shared(),
+                             &|| f.lock_shared()));
+            }
+            State::Unlocked => {}
+
+        }
+        Ok(FileLock { f: Some(f), path: path, state: state })
+    }
+}
+
+/// Acquires a lock on a file in a "nice" manner.
+///
+/// Almost all long-running blocking actions in Cargo have a status message
+/// associated with them as we're not sure how long they'll take. Whenever a
+/// conflicted file lock happens, this is the case (we're not sure when the lock
+/// will be released).
+///
+/// This function will acquire the lock on a `path`, printing out a nice message
+/// to the console if we have to wait for it. It will first attempt to use `try`
+/// to acquire a lock on the crate, and in the case of contention it will emit a
+/// status message based on `msg` to `config`'s shell, and then use `block` to
+/// block waiting to acquire a lock.
+///
+/// Returns an error if the lock could not be acquired or if any error other
+/// than a contention error happens.
+fn acquire(config: &Config,
+           msg: &str,
+           path: &Path,
+           try: &Fn() -> io::Result<()>,
+           block: &Fn() -> io::Result<()>) -> CargoResult<()> {
+    match try() {
+        Ok(()) => return Ok(()),
+        Err(e) => {
+            if e.raw_os_error() != lock_contended_error().raw_os_error() {
+                return Err(human(e)).chain_error(|| {
+                    human(format!("failed to lock file: {}", path.display()))
+                })
+            }
+        }
+    }
+    let msg = format!("waiting for file lock on {}", msg);
+    try!(config.shell().err().say_status("Blocking", &msg, CYAN));
+
+    block().chain_error(|| {
+        human(format!("failed to lock file: {}", path.display()))
+    })
+}
+
+fn create_dir_all(path: &Path) -> io::Result<()> {
```

Comment:
The version in the standard library doesn't currently handle the `NotFound` error as it's not intended to be run concurrently (which I have personally advocated for). This is basically the concurrent variant of those functions.

Maybe we should consider adding it to libstd at some point...